# Factorial

Zeros in 1000!

In [3]:
def trailing_zeros(n):
    count = 0
    while n >= 5:  # Keep going as long as n is at least 5
        n = n // 5  # Divide n by 5 (integer division, like floor)
        count += n  # Add the new n to the count
    return count

print(trailing_zeros(1000))  # Output

249


We can see there are 249 zeros in 1000!

# Big Numbers

Find the sum of the digits of (100^13 - 26)

In [4]:
#Calculate the number
x = 100**13 - 26

# Convert to string and split into digits
list_of_digits = [int(digit) for digit in str(x)]

# Sum the digits
digit_sum = sum(list_of_digits)

# Show results
print("Number:", x)
print("Digits:", list_of_digits)
print("Sum of digits:", digit_sum)


Number: 99999999999999999999999974
Digits: [9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 7, 4]
Sum of digits: 227


# Handshakes

In [8]:
team_size = 11
handshakes_between_teams = team_size * team_size

handshakes_within_winning_team = team_size * (team_size - 1) // 2

total_handshakes = handshakes_between_teams + handshakes_within_winning_team


print("Handshakes between teams:", handshakes_between_teams)
print("Handshakes within winning team:", handshakes_within_winning_team)
print("Total handshakes:", total_handshakes)

Handshakes between teams: 121
Handshakes within winning team: 55
Total handshakes: 176


# Harder Handshakers

In [11]:
def mr_smith_handshakes(n_couples: int) -> int:

    return n_couples - 1


print(mr_smith_handshakes(5))


4


At the party there are n couples, which makes 2n people in total. No one shakes hands with their own spouse, and no one shakes the same hand twice. When Mr. Smith asks the other (2n − 1) people how many hands they shook, he receives all different answers. The possible answers must be 0, 1, 2, …, up to (2n − 2). These answers come in pairs that always add up to (2n − 2), such as (0 and 2n−2), (1 and 2n−3), (2 and 2n−4), and so on. The only number that does not pair with another is the middle value, which is (n − 1). That number must belong to Mrs. Smith, and therefore Mr. Smith also shook (n − 1) hands.
